aiida/orm/utils/links.py

```diff
@@ -122,37 +122,59 @@ def validate_link(source, target, link_type, link_label):
     if not isinstance(source, type_source) or not isinstance(target, type_target):
         raise ValueError('cannot add a {} link from {} to {}'.format(link_type, type(source), type(target)))
 
-    # Validate the outdegree of the source node
-    outgoing = source.get_outgoing(link_type=link_type)
+    def link_triple_exists(source, target, link_type, link_label):
+        """Return whether a link with the given type and label exists between the given source and target node.
+
+        :param source: node from which the link is outgoing
+        :param target: node to which the link is incoming
+        :param link_type: the link type
+        :param link_label: the link label
+        :return: boolean, True if the link triple exists, False otherwise
+        """
+        from aiida.orm import QueryBuilder
+
+        if target._incoming_cache and LinkTriple(source, link_type, link_label) in target._incoming_cache:  # pylint: disable=protected-access
+            return True
+
+        builder = QueryBuilder()
+        builder.append(Node, filters={'uuid': source.uuid}, project=['id'])
```

Comment:
We can also use `id`s here, right? Queries on the ID are typically much faster (same for the target node one line below)

Reply:
We can. If you say it is faster I will change it

Reply:
In addition, the query part can be skipped if any of the nodes do not have a pk (if it is not stored).

Reply:
actually this is a good point. I think this might now except if one of the nodes is not stored since it won't have a pk. I guess we can revert to using UUIDs or just add `if source.pk is None or target.pk is None: return False`

Reply:
I would second the latter as it skips the need of querying the database. One of the major usage of this function is to validate the links of an unstored node just before having it stored.

Reply:
I have updated the function to include this check to potentially skip the query if either of the two nodes is not stored.